src/backend/vault-hsm/dh-curve25519.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2016 Southern Storm Software, Pty Ltd.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "internal.h"
+#include "crypto/ed25519/ed25519.h"
+#include <string.h>
+
+/* We use ed25519's faster curved25519_scalarmult_basepoint() function
+   when deriving a public key from a private key.  Unfortunately ed25519
+   doesn't have an equivalent function for general curve25519 calculations
+   so we fall back to the curve25519-donna implementation for that. */
+
+int curve25519_donna(uint8_t *mypublic, const uint8_t *secret, const uint8_t *basepoint);
+
+typedef struct
+{
+    struct NoiseDHState_s parent;
+    uint8_t private_key[32];
+    uint8_t public_key[32];
+
+} NoiseCurve25519State;
+
+static int noise_curve25519_generate_keypair
+    (NoiseDHState *state, const NoiseDHState *other)
```

Comment:
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?